pkg/library/dag/dag.go

```diff
@@ -0,0 +1,283 @@
+package dag
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/kuadrant/kuadrant-operator/pkg/library/utils"
+)
+
+// A Directed Acyclic Graph (DAG) is a graph representing a structure formed by vertices, or nodes,
+// connected by directed edges.
+// In a DAG, each edge has an initial node, called the parent, and a final node, called the child.
+// The graph is considered acyclic because it does not contain any cycles,
+// meaning there are no sequences of consecutive directed edges that form a closed loop.
+
+type NodeID = string
+
+type nodeNotFoundError struct {
+	id NodeID
+}
+
+func (e *nodeNotFoundError) Error() string {
+	return fmt.Sprintf("node %s not found", e.id)
+}
+
+func IsNodeNotFound(err error) bool {
+	var nodeNotFoundErr *nodeNotFoundError
+	return errors.As(err, &nodeNotFoundErr)
+}
+
+type Node interface {
+	ID() NodeID
+}
+
+type internalNode struct {
+	id       NodeID
+	node     Node
+	parents  map[NodeID]*internalNode
+	children map[NodeID]*internalNode
+}
+
+type Options struct {
+	fieldIndexers []FieldIndexer
+}
+
+type NodeLabel = string
+type Field = string
+
+type IndexerFunc func(Node) []NodeLabel
+
+func WithFieldIndexer(f Field, e IndexerFunc) *FieldIndexer {
+	return &FieldIndexer{f, e}
+}
+
+type FieldIndexer struct {
+	field   Field
+	indexer IndexerFunc
+}
+
+func (f FieldIndexer) ApplyTo(opts *Options) {
+	opts.fieldIndexers = append(opts.fieldIndexers, f)
+}
+
+type Opt interface {
+	// ApplyTo applies this configuration to the given options.
+	ApplyTo(*Options)
+}
+
+var _ Opt = FieldIndexer{}
+
+type DAG struct {
+	nodes         map[NodeID]*internalNode
+	fieldIndexers []FieldIndexer
+	nodeIndexes   map[Field]map[NodeLabel][]Node
+}
+
+func NewDAG(opts ...Opt) *DAG {
+	// Capture options
+	dagOpts := &Options{}
+	for _, opt := range opts {
+		opt.ApplyTo(dagOpts)
+	}
+
+	return &DAG{
+		nodes:         make(map[NodeID]*internalNode),
+		fieldIndexers: dagOpts.fieldIndexers,
+		nodeIndexes:   make(map[Field]map[NodeLabel][]Node),
+	}
+}
+
+func (d *DAG) AddNode(node Node) error {
+	n := &internalNode{
+		id:       node.ID(),
+		node:     node,
+		parents:  make(map[string]*internalNode),
+		children: make(map[string]*internalNode),
+	}
+
+	if _, exists := d.nodes[n.id]; exists {
+		return fmt.Errorf("node %s already exists", n.id)
+	}
+
+	d.nodes[n.id] = n
+	d.populateIndexes(node)
```

Comment:
We should add a mutex and wrap this as a critical section.

Reply:
I am aware that the dag package is not thread safe. But, as it is used now, it does not need to be thread-safe. Locks can be added when (if) a single instance is accessed by multiple threads.